File: Back-End/IntelliView/IntelliView.DataAccess/Services/AvatarService.cs
namespace IntelliView.DataAccess.Services
{
    using IntelliView.DataAccess.Services.IService;
    using IntelliView.Models.Models;
    using Microsoft.Extensions.Logging;
    using Newtonsoft.Json;
    using Newtonsoft.Json.Linq;
    using System;
    using System.Net.Http;
    using System.Text;
    using System.Threading.Tasks;


    public class AvatarService : IAvatarService
    {
        private static readonly HttpClient client = new HttpClient();
        private const string SubscriptionKey = "306a1f1015aa4cf6b54d6ab9c8032559";
        private const string ServiceRegion = "westeurope";
        private const string ServiceHost = "customvoice.api.speech.microsoft.com";

        private readonly ILogger<AvatarService> _logger; // Added logger field

        public AvatarService(ILogger<AvatarService> logger) // Added logger parameter in constructor
        {
            _logger = logger;
        }

        public async Task<((string downloadUrl, dynamic VideoId)? video, string message)> SubmitSynthesis(
          string displayName = "Simple avatar synthesis",
          string description = "Simple avatar synthesis description",
          string textType = "PlainText",
          string voice = "en-US-JennyNeural",
          string inputText = "Hi, This is intelliview Ai service and I will be with you for the Rest of this interview",
          bool customized = false,
          string talkingAvatarCharacter = "lisa",
          string talkingAvatarStyle = "graceful-sitting",
          string videoFormat = "webm",
          string videoCodec = "vp9",
          string subtitleType = "soft_embedded",
          string backgroundColor = "transparent")
        {
            var url = $"https://{ServiceRegion}.{ServiceHost}/api/texttospeech/3.1-preview1/batchsynthesis/talkingavatar";
            client.DefaultRequestHeaders.Add("Ocp-Apim-Subscription-Key", SubscriptionKey);

            var payload = new
            {
                displayName,
                description,
                textType,
                synthesisConfig = new { voice },
                customVoices = new { },
                inputs = new[] { new { text = inputText } },
                properties = new
                {
                    customized,
                    talkingAvatarCharacter,
                    talkingAvatarStyle,
                    videoFormat,
                    videoCodec,
                    subtitleType,
                    backgroundColor
                }
            };

            var content = new StringContent(JsonConvert.SerializeObject(payload), Encoding.UTF8, "application/json");

            var response = await client.PostAsync(url, content);
            try
            {
                if (response.IsSuccessStatusCode)
                {
                    var responseJson = JsonConvert.DeserializeObject<dynamic>(await response.Content.ReadAsStringAsync());
                    var jobId = responseJson?.id.ToString();
                    int timer = 0;
                    string downloadUrl;
                    do
                    {
                        await Task.Delay(5000); // wait for 5 seconds before checking the status
                        downloadUrl = await GetSynthesis(jobId);
                        timer++;
                        if (timer == 15) break;
                    }
                    while (downloadUrl == null);

                    return timer == 15 ? (null, "Generating Video took too long") : ((downloadUrl, jobId!), "successful");
                }
            }
            catch (Exception e)
            {
                _logger.LogError(e, "Failed to submit batch avatar synthesis job"); // Added logger error message

                return (null, message: $"Failed to submit batch avatar synthesis job: {response.StatusCode}, \n error: {e.Message}");
            }

            _logger.LogError("Failed to get batch synthesis job"); // Added logger error message
            return (null, $"Failed to get batch synthesis job: {response.StatusCode}");
        }

        public async Task<string?> GetSynthesis(string jobId)
        {
            var url = $"https://{ServiceRegion}.{ServiceHost}/api/texttospeech/3.1-preview1/batchsynthesis/talkingavatar/{jobId}";
            client.DefaultRequestHeaders.Add("Ocp-Apim-Subscription-Key", SubscriptionKey);

            var response = await client.GetAsync(url);
            try
            {
                if (response.IsSuccessStatusCode)
                {
                    var responseJson = JsonConvert.DeserializeObject<dynamic>(await response.Content.ReadAsStringAsync());
                    if (responseJson?.status == "Succeeded")
                    {
                        _logger.LogInformation("Successfully retrieved batch synthesis job"); // Added logger information message
                        return responseJson.outputs.result;
                    }
                    _logger.LogInformation("Batch synthesis job status is not 'Succeeded'"); // Added logger information message
                    return null;
                }
            }
            catch (Exception e)
            {
                _logger.LogError(e, "Failed to get batch synthesis job"); // Added logger error message

                throw new Exception($"Failed to get batch synthesis job: {response.StatusCode}, \n error: {e.Message} ");
            }

            throw new Exception($"Failed to get batch synthesis job: {response.StatusCode}");
        }

        public async Task<List<dynamic>> ListSynthesisJobs(int skip = 0, int top = 100)
        {
            var url = $"https://{ServiceRegion}.{ServiceHost}/api/texttospeech/3.1-preview1/batchsynthesis/talkingavatar?skip={skip}&top={top}";
            client.DefaultRequestHeaders.Add("Ocp-Apim-Subscription-Key", SubscriptionKey);

            var response = await client.GetAsync(url);

            if (response.IsSuccessStatusCode)
            {
                var responseJson = JsonConvert.DeserializeObject<dynamic>(await response.Content.ReadAsStringAsync());
                _logger.LogInformation($"List batch synthesis jobs successfully, got {responseJson?.values.Count} jobs");
                return ((JArray)responseJson.values).ToObject<List<dynamic>>();
            }
            else
            {
                _logger.LogError($"Failed to list batch synthesis jobs: {response.StatusCode}"); // Added logger error message
                throw new Exception($"Failed to list batch synthesis jobs: {response.StatusCode}");
            }
        }

        // voices https://learn.microsoft.com/en-us/azure/ai-services/speech-service/language-support?tabs=tts
        public string getAvatarVoice(MockLang lang)
        {
            switch (lang)
            {
                case MockLang.English:
                    return "en-GB-SoniaNeural";
                case MockLang.Arabic:
                    return "ar-YE-MaryamNeural";
                default:
                    return "en-IE-EmilyNeural";
            }
        }
    }
}

File: Back-End/IntelliView/IntelliView.DataAccess/Services/avatarService.py
#!/usr/bin/env python
# coding: utf-8

# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE.md file in the project root for full license information.

import json
import logging
import os
import sys
import time
from pathlib import Path
import argparse
import requests

logging.basicConfig(stream=sys.stdout, level=logging.INFO,  # set to logging.DEBUG for verbose output
        format="[%(asctime)s] %(message)s", datefmt="%m/%d/%Y %I:%M:%S %p %Z")
logger = logging.getLogger(__name__)

# Your Speech resource key and region
# This example requires environment variables named "SPEECH_KEY" and "SPEECH_REGION"
parser = argparse.ArgumentParser()
parser.add_argument("--voice", default="en-US-JennyNeural")
parser.add_argument("--name", default="Simple avatar synthesis")
parser.add_argument("--description", default="Simple avatar synthesis description")
parser.add_argument("--input", default="Hi, I'm a virtual assistant created by Microsoft.")
parser.add_argument("--service_host", default="customvoice.api.speech.microsoft.com")
parser.add_argument("--talking_avatar_character", default="lisa")
parser.add_argument("--talking_avatar_style", default="graceful-sitting")
parser.add_argument("--background_color", default="transparent")
parser.add_argument("--subtitle_type", default="soft_embedded")
parser.add_argument("--videocodec", default="vp9")
parser.add_argument("--text_type", default="PlainText")

SUBSCRIPTION_KEY = os.getenv("SUBSCRIPTION_KEY", '306a1f1015aa4cf6b54d6ab9c8032559')
SERVICE_REGION = os.getenv("SERVICE_REGION", "westeurope")

args = parser.parse_args()
VOICE = args.voice
NAME = args.name
DESCRIPTION = args.description
INPUT = args.input
SERVICE_HOST = args.service_host
TALKING_AVATAR_CHARACTER = args.talking_avatar_character
TALKING_AVATAR_STYLE = args.talking_avatar_style
BACKGROUND_COLOR = args.background_color
SUBTITLE_TYPE = args.subtitle_type
VIDEOCODEC = args.videocodec
TEXT_TYPE = args.text_type
def submit_synthesis():
    url = f'https://{SERVICE_REGION}.{SERVICE_HOST}/api/texttospeech/3.1-preview1/batchsynthesis/talkingavatar'
    header = {
        'Ocp-Apim-Subscription-Key': SUBSCRIPTION_KEY,
        'Content-Type': 'application/json'
    }

    payload = {
        'displayName': NAME,
        'description': DESCRIPTION,
        "textType": TEXT_TYPE,
        'synthesisConfig': {
            "voice": VOICE,
        },
        # Replace with your custom voice name and deployment ID if you want to use custom voice.
        # Multiple voices are supported, the mixture of custom voices and platform voices is allowed.
        # Invalid voice name or deployment ID will be rejected.
        'customVoices': {
            # "YOUR_CUSTOM_VOICE_NAME": "YOUR_CUSTOM_VOICE_ID"
        },
        "inputs": [
            {
                "text": INPUT,
            },
        ],
        "properties": {
            "customized": False, # set to True if you want to use customized avatar
            "talkingAvatarCharacter": TALKING_AVATAR_CHARACTER,  # talking avatar character
            "talkingAvatarStyle": TALKING_AVATAR_STYLE,  # talking avatar style, required for prebuilt avatar, optional for custom avatar
            "videoFormat": "webm",  # mp4 or webm, webm is required for transparent background
            "videoCodec": VIDEOCODEC,  # hevc, h264 or vp9, vp9 is required for transparent background; default is hevc
            "subtitleType": SUBTITLE_TYPE,
            "backgroundColor": BACKGROUND_COLOR,
        }
    }

    response = requests.post(url, json.dumps(payload), headers=header)
    if response.status_code < 400:
        logger.info('Batch avatar synthesis job submitted successfully')
        logger.info(f'Job ID: {response.json()["id"]}')
        return response.json()["id"]
    else:
        logger.error(f'Failed to submit batch avatar synthesis job: {response.text}')


def get_synthesis(job_id):
    url = f'https://{SERVICE_REGION}.{SERVICE_HOST}/api/texttospeech/3.1-preview1/batchsynthesis/talkingavatar/{job_id}'
    header = {
        'Ocp-Apim-Subscription-Key': SUBSCRIPTION_KEY
    }
    response = requests.get(url, headers=header)
    if response.status_code < 400:
        logger.debug('Get batch synthesis job successfully')
        logger.debug(response.json())
        if response.json()['status'] == 'Succeeded':
            logger.info(f'Batch synthesis job succeeded, download URL: {response.json()["outputs"]["result"]}')
        return response.json()['status']
    else:
        logger.error(f'Failed to get batch synthesis job: {response.text}')
  
  
def list_synthesis_jobs(skip: int = 0, top: int = 100):
    """List all batch synthesis jobs in the subscription"""
    url = f'https://{SERVICE_REGION}.{SERVICE_HOST}/api/texttospeech/3.1-preview1/batchsynthesis/talkingavatar?skip={skip}&top={top}'
    header = {
        'Ocp-Apim-Subscription-Key': SUBSCRIPTION_KEY
    }
    response = requests.get(url, headers=header)
    if response.status_code < 400:
        logger.info(f'List batch synthesis jobs successfully, got {len(response.json()["values"])} jobs')
        logger.info(response.json())
    else:
        logger.error(f'Failed to list batch synthesis jobs: {response.text}')
  
  
if __name__ == '__main__':
    job_id = submit_synthesis()
    if job_id is not None:
        while True:
            status = get_synthesis(job_id)
            if status == 'Succeeded':
                logger.info('batch avatar synthesis job succeeded')
                break
            elif status == 'Failed':
                logger.error('batch avatar synthesis job failed')
                break
            else:
                logger.info(f'batch avatar synthesis job is still running, status [{status}]')
                time.sleep(5)
